### PR TITLE
Alerting: Introduce proper feature toggles for common state history backend combinations

### PR DESCRIPTION
**Pull request by alexweav**
**What is this feature?**

Some operators prefer to control Grafana using the proper feature toggle system rather than plain configuration.

This PR adds 3 first-class feature toggles for state history, that represent a typical progression of enablement:
1. Toggle 1 allows Loki as a secondary backend
2. Toggle 2 allows Loki to be primarily used for queries, with backup writes to annotations
3. Toggle 3 allows full control of the backend.

Depending on the level of restrictiveness chosen, backend types may be coerced to the closest allowable thing. When multiple toggles are disabled, we respect the most restrictive one.

**Why do we need this feature?**

Helps operators test out the system better, as a staged rollout.

Grafana now ships with smarter presets for "suggested" configurations.

**Who is this feature for?**

Grafana Operators wishing to configure Loki-based state history.

**Which issue(s) does this PR fix?**:

n/a

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
- [x] There are no known compatibility issues with older supported versions of Grafana, or plugins.
- [x] It passes the [Hosted Grafana feature readiness review](https://docs.google.com/document/d/1QL9Ly8KnXzpb6ISbg49pTODRO5mhA5tkkfIZVX6pqQU/edit) for observability, scalability, performance, and security.

### CODE REVIEW

File: pkg/setting/setting_unified_alerting.go [hunk 1]
```diff
@@ -137,7 +137,7 @@ func (cfg *Cfg) readUnifiedAlertingEnabledSetting(section *ini.Section) (*bool,
 	// than disable it. This issue can be found here
 	hasEnabled := section.Key("enabled").Value() != ""
 	if !hasEnabled {
-		// TODO: Remove in Grafana v9
+		// TODO: Remove in Grafana v10
```
Comment (alexweav) [inline pkg/setting/setting_unified_alerting.go:140]: :sob:

File: docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md [hunk 2]
```diff
@@ -99,6 +99,9 @@ Alpha features might be changed or removed without prior notice.
 | `influxdbBackendMigration`         | Query InfluxDB InfluxQL without the proxy                                                                                                                                                           |
 | `clientTokenRotation`              | Replaces the current in-request token rotation so that the client initiates the rotation                                                                                                            |
 | `prometheusDataplane`              | Changes responses to from Prometheus to be compliant with the dataplane specification. In particular it sets the numeric Field.Name from 'Value' to the value of the `__name__` label when present. |
+| `alertStateHistoryLokiSecondary`   | Enable alert state history to be written to an external loki instance in addition to Grafana annotations.                                                                                           |
+| `alertStateHistoryLokiPrimary`     | Enable usage of a remote Loki instance as the primary source for state history reads.                                                                                                               |
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:103]:
```suggestion
| `alertStateHistoryLokiPrimary`     | Enable a remote Loki instance to be the primary source for state history reads.                                                                                                               |
```

File: docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md [hunk 3]
```diff
@@ -99,6 +99,9 @@ Alpha features might be changed or removed without prior notice.
...
+| `alertStateHistoryLokiOnly`        | Allow Grafana annotations emitted by alerting to be disabled completely when remote Loki is in place.                                                                                               |
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:104]:
```suggestion
| `alertStateHistoryLokiOnly`        | Disable Grafana alerts from emitting annotations when a remote Loki instance is available.                                                                                               |
```

**Review by chri2547 — CHANGES_REQUESTED** [review]
A few suggestions and a question. Thank you!

File: docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md [hunk 4]
```diff
@@ -99,6 +99,9 @@ Alpha features might be changed or removed without prior notice.
 | `influxdbBackendMigration`         | Query InfluxDB InfluxQL without the proxy                                                                                                                                                           |
 | `clientTokenRotation`              | Replaces the current in-request token rotation so that the client initiates the rotation                                                                                                            |
 | `prometheusDataplane`              | Changes responses to from Prometheus to be compliant with the dataplane specification. In particular it sets the numeric Field.Name from 'Value' to the value of the `__name__` label when present. |
+| `alertStateHistoryLokiSecondary`   | Enable alert state history to be written to an external loki instance in addition to Grafana annotations.                                                                                           |
```
[the remark this answers is not in the document]
Comment (alexweav) [reply docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:102]: Sure, right now alerting happens to (ab)use Grafana Annotations as a method of storing history for when alerts fire/resolve. It writes a lot of rows to the annotations table that aren't really associated with "lines on panels" which was the intended purpose of annotations.

We're moving the storage mechanism for this data off annotations to something more bespoke and giving alerting its own API for state history. The new storage has pluggable backends (including annotations, if OSS customers wish to keep it), but in grafana cloud, we'll be automatically configuring a Loki instance for users as the default store for this data. Queries will dispatch to whichever store is configured.

The idea is to have three stages of maturity here, for users who want to try enabling the new Loki backend:
- One where we write to loki in addition to the current SQL table (which only _happens_ to be the annotations table)
- One where Loki will be the primary backend for serving queries, but we still write the old annotations as a backup
- One where we allow a full switch-over to the external system

File: docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md (hunk 3 above)
Comment (alexweav) [reply docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:104]: This is much better wording, great suggestion

**Review by gotjosh — APPROVED** [review]
LGTM

File: docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md]: 👍

**Review by chri2547 — APPROVED** [review]
Docs LGTM! Thank you!

File: pkg/services/ngalert/ngalert.go [hunk 5]
```diff
@@ -438,3 +441,48 @@ func configureHistorianBackend(ctx context.Context, cfg setting.UnifiedAlertingS
...
+			logger.Info("Coercing Loki to a secondary backend due to state history feature toggles")
+			cfg.MultiPrimary = "annotations"
+			cfg.MultiSecondaries = []string{"loki"}
+		}
+		// If we're using loki, we are only allowed to use it as a secondary. Dual write to it, plus annotations.
+		if backend == historian.BackendTypeLoki {
+			logger.Info("Coercing Loki to dual writes with a secondary backend due to state history feature toggles")
+			cfg.Backend = "multiple"
+			cfg.MultiPrimary = "annotations"
+			cfg.MultiSecondaries = []string{"loki"}
```
Comment (santihernandezc) [inline pkg/services/ngalert/ngalert.go:474]: The only difference between this `if` condition and the previous one is the log message. The effect is the same:
- Backend type is "multiple".
- Primary backend is "annotations".
- Secondaries is a slice containing "loki".

Maybe we can de-duplicate this code. Also, there's no need to check both conditions, they are mutually exclusive.

Comment (santihernandezc) [reply pkg/services/ngalert/ngalert.go:474]: This change is not strictly necessary, you can leave it as it is for readability reasons.

**Review by santihernandezc — CHANGES_REQUESTED** [review]
I left a few comments, my next review will contain results from manual testing.

File: pkg/services/featuremgmt/registry.go [hunk 6]
```diff
@@ -460,5 +460,23 @@ var (
 			State:       FeatureStateAlpha,
 			Owner:       grafanaObservabilityMetricsSquad,
 		},
+		{
+			Name:        "alertStateHistoryLokiSecondary",
+			Description: "Enable alert state history to be written to an external loki instance in addition to Grafana annotations.",
```
[the remark this answers is not in the document]
Comment (alexweav) [reply pkg/services/featuremgmt/registry.go]: True, but if this particular toggle is off, then it won't even bother checking the value of the other toggles. I.e. this toggle must be enabled to allow Grafana to write to an external Loki instance.

File: pkg/services/ngalert/ngalert.go [hunk 7]
```diff
@@ -438,3 +441,48 @@ func configureHistorianBackend(ctx context.Context, cfg setting.UnifiedAlertingS
...
 }
+
+// applyStateHistoryFeatureToggles edits state history configuration to comply with currently active feature toggles.
+func applyStateHistoryFeatureToggles(cfg *setting.UnifiedAlertingStateHistorySettings, ft featuremgmt.FeatureToggles, logger log.Logger) {
+	backend, _ := historian.ParseBackendType(cfg.Backend)
+	// These feature toggles represent specific, common backend configurations.
+	// If all toggles are enabled, we listen to the state history config as written.
+	// If any of them are disabled, we ignore the configured backend and treat the toggles as an override.
+	// If multiple toggles are disabled, we go with the most "restrictive" one.
+	if !ft.IsEnabled("alertStateHistoryLokiSecondary") {
```
[the remark this answers is not in the document]
Comment (alexweav) [reply pkg/services/ngalert/ngalert.go:452]: I wrote this before I ran the code generator to produce said constants :D So I didn't realize that the constants existed. Good call!

File: pkg/services/featuremgmt/registry.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (santihernandezc) [reply pkg/services/featuremgmt/registry.go]: I mean, having the same description we have in our docs but in our code 😬

**Review by santihernandezc — APPROVED** [review]
I tried the different toggles, querying both Loki and the `annotation` table on each test. Everything seems to work as expected.

File: pkg/services/ngalert/ngalert.go [hunk 8]
```diff
@@ -438,3 +441,48 @@ func configureHistorianBackend(ctx context.Context, cfg setting.UnifiedAlertingS
...
+		// If we cannot even treat Loki as a secondary, we must use annotations only.
+		if backend == historian.BackendTypeMultiple || backend == historian.BackendTypeLoki {
+			logger.Info("Forcing Annotation backend due to state history feature toggles")
+			cfg.Backend = "annotations"
```
[the remark this answers is not in the document]
Comment (alexweav) [reply pkg/services/ngalert/ngalert.go:456]: Done! I originally didn't do this as this all runs before we actually parse the backend type for real. But you're right, this is a little more change-proof.

**Comment by jdbaldry** [comment]
@alexweav, should this be backported into v9.5.x?

**Comment by alexweav** [comment]
This PR was merged before the 9.5 cutoff... and this code change is present in the v9.5.x branch already: https://github.com/grafana/grafana/blob/f26e56bc43477dde23fccccea60b618b95aebe6c/pkg/services/ngalert/ngalert.go#L443

So I don't think it needs to be @jdbaldry? Unless there's something I am missing?

**Comment by jdbaldry** [comment]
Cool, thanks for checking. I think that there was just some confusion about the milestone.

@zerok Is the correct action here to update the milestone back to `9.4.x`?

**Pull request merged**